src/rgw/rgw_op.cc

```diff
@@ -3403,6 +3553,34 @@ void RGWDeleteObj::execute()
       if (op_ret >= 0) {
         delete_marker = del_op.result.delete_marker;
         version_id = del_op.result.version_id;
+        map<string, bufferlist>::iterator cmp = s->bucket_attrs.find(RGW_ATTR_COMPRESSION);
+        if (cmp != s->bucket_attrs.end()) {
+          RGWBucketCompressionInfo bucket_size;
+          uint64_t deleted_size;
+          try {
+            ::decode(bucket_size, cmp->second);
+            bool tmp;
+            RGWCompressionInfo cs_info;
+            op_ret = rgw_compression_info_from_attrset(attrs, tmp, cs_info);
+            if (op_ret < 0) {
+              lderr(s->cct) << "ERROR: failed to decode compression info, cannot update stat" << dendl;
+              deleted_size = s->obj_size;
+            } else {
+              if (tmp) 
+                deleted_size = cs_info.orig_size;
+              else
+                deleted_size = s->obj_size;
+            }
+            bucket_size.orig_size -= deleted_size;
+            bufferlist bs;
+            ::encode(bucket_size, bs);
+            s->bucket_attrs[RGW_ATTR_COMPRESSION] = bs;
+            op_ret = store->put_bucket_instance_info(s->bucket_info, false, real_time(),
+              &s->bucket_attrs);
+          } catch (buffer::error& err) {
+            lderr(s->cct) << "ERROR: failed to read decompressed bucket size, cannot update stat" << dendl;
+          }          
+        }
```

Comment:
can we move this into a helper function? adding some early returns could simplify things and cut down on nesting. for example, if the catch {} returns an error, the try {} only has to cover the ::decode() line

that would also make it easier to reason about error handling; if `del_op.delete_obj()` has already succeeded, i'm not sure we should fail the op if we hit an error updating the bucket instance. something like this?
```diff
 if (op_ret >= 0) {
   delete_marker = del_op.result.delete_marker;
   version_id = del_op.result.version_id;
+
+  int ret = update_compressed_bucket_size(...);
+  if (ret < 0) {
+    // print an error message but don't set op_ret
+  }
```